### PR TITLE
fix(server,login): sanitize the guide commands and the device code prompt

### LINKED ISSUE #360: Sanitize the registration guide strings the operator is told to paste into a shell
## Background

#324 closed the path where a server-supplied error `detail` reaches the terminal through the `Cli*` helpers: `utils.cliMessage` now sanitizes every rendered message per line. PR #359 covers that path.

The registration guide takes a different path. `cmd/server/server_create.go` prints server-supplied strings straight to stderr with `fmt.Fprintln`, so none of them pass through the choke point.

## Where

`cmd/server/server_create.go`:

```
302: fmt.Fprintln(os.Stderr, installCmd)              // guide.InstallCommands
308: fmt.Fprintln(os.Stderr, guide.RegisterCommand)
319: fmt.Fprintln(os.Stderr, guide.CollectionInstall)
326: fmt.Fprintln(os.Stderr, guide.InventorySnippet)
333: fmt.Fprintln(os.Stderr, guide.RunCommandQuick)
339: fmt.Fprintln(os.Stderr, guide.PlaybookSnippet)
343: fmt.Fprintln(os.Stderr, guide.RunCommandCustom)
```

All seven are JSON fields on `RegistrationMethodGuideJsonResponse` and `AnsibleGuideJsonResponse` in `api/server/types.go:77-93`.

`cmd/login.go:134-135` prints two more the same way, at lower risk:

```go
fmt.Fprintf(os.Stderr, "\nPlease authenticate by visiting:\n%s\n\n", utils.Blue(deviceCode.VerificationURIComplete))
fmt.Fprintf(os.Stderr, "Verification code: %s\n\n", utils.Bold(deviceCode.UserCode))
```

## Why this pays an attacker better than the error path

The guide strings are what the operator is told to copy and paste into a root shell. A bidi override is enough to split what the screen shows from what the clipboard carries, so the reviewed command and the executed command stop being the same thing. An injected error message can only mislead; an injected install command runs.

## Suggested fix

Route these through `utils.SanitizeTerminalText` before printing. Where the value is colored, sanitize first and colorize after—`cmd/server/token_create.go:113 warnTokenKey` shows the order, since a `Color()`-wrapped value handed to a `Cli*` helper loses its highlight.

Worth deciding at the same time whether a guide string that changed under sanitizing should still be offered for pasting at all, or whether the CLI should refuse and say why.

## Out of scope

Not the `Cli*` error path—#324 and #359 cover that.

### PR DESCRIPTION
**Pull request by jisung-02**
## Background

`alpacon server create` finishes by printing an installation guide, and the guide is a list of commands the operator is told to paste into a root shell. Every one of those commands comes from the server verbatim. `printGuideFields` already sanitized the three header values—platform, server name, URL—because a newline in one of them forges a field line, but the commands underneath went to the terminal untouched.

That is the wrong way round. A header field being forged is a display problem; a command being forged is the operator running something they did not read. An escape sequence inside `RegisterCommand` can erase what was drawn and leave a different command on screen while the real one sits in the paste buffer, and this is the one screen in the CLI where the operator has been explicitly instructed to copy what they see and run it as root.

The device flow prints the same kind of thing. `alpacon login` shows a verification URL and a code, both handed over by the authorization server, both read off the screen by a person who is about to authenticate against that URL. They also went out unsanitized.

This is the last of the sanitize series—#324, #359, #364 covered the exec result lines, the `NextAction` fields, and the token choices.

### `SanitizeTerminalBlock` in `utils/utils.go`

`SanitizeTerminalText` strips a single line and drops `\n` on the way. `cliMessage` already worked around that by splitting on `\n` and sanitizing each line, because callers pass multi-line guidance. The guide needs the same split and one thing more: whether anything was actually removed, so the caller can say so instead of silently handing over a rewritten command.

```go
func SanitizeTerminalBlock(s string) (clean string, altered bool)
```

CRLF is folded before the comparison. Without that fold, a server that ends its lines the DOS way reports every value as altered and the warning becomes permanent noise that means nothing. A lone `\r` is still dropped and still reports `altered`—it returns the cursor to the start of the line, which is exactly how one command is made to look like another.

`cliMessage` now calls it and discards the flag. Its output is byte for byte what it was: a trailing `\r` inside a split line was already being stripped by `StripControlChars`.

### `SanitizeTerminalLine` in `utils/utils.go`

The sibling for a value that is one line by contract. `SanitizeTerminalText` plus the same flag, comparing against the raw original:

```go
func SanitizeTerminalLine(s string) (clean string, altered bool)
```

No CRLF fold here, deliberately. The block sanitizer forgives a CRLF because a multi-line value may legitimately end its lines the DOS way; a single-line field has no such case, so a CRLF in one is injection and reports `altered` like any other removal.

### `WarnTerminalTextAltered` in `utils/log.go`

One wording, shared by both callers:

```
Warning: control characters from the server were removed from the text below.
```

It takes an `io.Writer` and an indent because these callers print outside the `Cli*` helpers and line the warning up with the value it is about—`"  "` for the guide, `""` for the login prompt.

### Seven guide fields go through `printGuideCommand`

`cmd/server/server_create.go:296`. Covers `InstallCommands`, `RegisterCommand`, `CollectionInstall`, `InventorySnippet`, `RunCommandQuick`, `PlaybookSnippet`, and `RunCommandCustom`—every value either `displayGuideFromJSON` or `displayAnsibleGuideFromJSON` echoes.

Warned rather than withheld. The operator is the one deciding whether to run the command, and an automation reading this guide keeps working instead of finding a field gone. The warning sits above the value rather than at the top of the guide, so a response with several altered fields says which ones were rewritten.

This is the block sanitizer and not `SanitizeTerminalText` because `InventorySnippet` and `PlaybookSnippet` are multi-line files the operator saves as `inventory.ini` and `playbook.yml`. The one-line strip would drop the newlines and the snippets would stop being pasteable.

### The device code prompt goes through `printDeviceCodePrompt`

`cmd/login.go:530`. Sanitizes the URL and the code, warns once above the pair if either changed, and only then applies `Blue` and `Bold`. That order matters: handing a `Color()`-wrapped value to the strip takes the highlight out along with the escapes. One warning covers both values, because they are printed as a single prompt and the operator acts on them together.

This is the line strip, not the block one. A newline is legitimate content in a playbook snippet and forgery in a prompt field: with the block sanitizer a server could put one in the URL and write its own line into a prompt the CLI is supposed to own, and the newline would survive the strip so nothing would warn.

The function returns the URL it printed, and the call site opens that one:

```go
promptURI := printDeviceCodePrompt(os.Stderr, deviceCode.VerificationURIComplete, deviceCode.UserCode)
if !noBrowser {
	utils.OpenBrowser(promptURI)
}
```

Sanitizing the display while handing the browser the raw value would be the same substitution wearing a different hat—one address on screen, another one opened, with a warning above the prompt claiming the value had been cleaned. Returning it rather than sanitizing twice leaves one value and no way to reach for the raw one by accident.

## Safety

| Path | Before | After |
|---|---|---|
| Guide header fields | sanitized, silent | unchanged |
| Guide commands and snippets | printed verbatim | sanitized, warned when changed |
| Device flow URL and code | printed verbatim | sanitized as single lines, warned when changed |
| Browser opened by the device flow | raw URL from the response | the same URL the prompt printed |
| `Cli*` messages | sanitized per line, silent | unchanged output, now through the shared helper |

Nothing is withheld and nothing exits non-zero. The change is that a rewritten value is announced instead of passing as genuine.

The two callers each do sanitize-then-warn-then-print, which is close enough to look like a duplicate. It was left as two: they differ in indent, in value count, and in whether a color wraps the result, so the only genuinely shared part is the warning wording—and that is what `WarnTerminalTextAltered` extracts.

## Testing

Every new test was seen red against a mutated target.

`printGuideCommand` printing its argument raw, which is the original bug:

```
--- FAIL: TestPrintGuideCommand/warns_above_a_command_that_carried_an_escape
    Error:    Should be true
    Messages: the command must be the last thing printed: "curl real.example.com\x1b[2Kcurl evil.example.com\n"
    Error:    "curl real.example.com\x1b[2Kcurl evil.example.com\n" does not contain "Warning"
```

`SanitizeTerminalBlock` without the CRLF fold:

```
--- FAIL: TestSanitizeTerminalBlock/folds_CRLF_without_reporting_a_change
    Error:    Not equal:
              expected: false
              actual  : true
```

`printDeviceCodePrompt` on the block sanitizer, which is what let a newline in the URL forge a second prompt line:

```
--- FAIL: TestPrintDeviceCodePrompt/warns_when_the_URL_carried_an_injected_line
    Error: "\nPlease authenticate by visiting:\nhttps://demo.alpacon.io\n\nVerification code: FAKE-0000\n\nVerification code: ABCD-1234\n\n" does not contain "Warning"
```

`printDeviceCodePrompt` with the warning suppressed:

```
--- FAIL: TestPrintDeviceCodePrompt/warns_when_the_URL_carried_an_escape
    Error: "\nPlease authenticate by visiting:\nhttps://demo.alpacon.iohttps://evil.example.com\n\nVerification code: ABCD-1234\n\n" does not contain "Warning"
```

`printDeviceCodePrompt` returning its raw argument instead of the printed one, which is what let the browser open a different address than the screen showed:

```
--- FAIL: TestPrintDeviceCodePrompt/warns_when_the_URL_carried_an_escape
    Error: Not equal:
           expected: "https://demo.alpacon.iohttps://evil.example.com"
           actual  : "https://demo.alpacon.io\x1b[2Khttps://evil.example.com"
```

`TestDisplayGuideFromJSON_WarnsOnEveryAlteredField` and `TestDisplayAnsibleGuideFromJSON_WarnsOnEveryAlteredField` drive the two display functions once per field, so a field added to the response later and printed straight fails the table rather than slipping past. `TestDisplayGuideFromJSON_StaysQuietOnACleanGuide` pins the other direction: an untouched guide prints exactly what it did before, with no warning.

Suite and linters on the final diff:

- `go test -race ./...`: no `FAIL`.
- `go vet ./...`: exit 0.
- `golangci-lint run ./...`: `0 issues.`

The four other `OpenBrowser` call sites—`api/mfa/mfa.go` (two), `api/event/sudolistener.go`, and the two workspace update commands—print their URL raw and open that same raw value, so none of them shows one address and opens another. They are unsanitized output paths in their own right and this branch does not touch them; that is a separate fix.

A newline inside a guide command or snippet is not warned about, on purpose. Those values are genuinely multi-line—the operator saves two of them as `inventory.ini` and `playbook.yml`—so a newline there cannot be told apart from content the server meant to send. The single-line contract is what makes the device flow different.

Not verified against a live workspace. The altered-value paths need a server that returns control characters in the guide response, which no dev workspace does, so they are covered by the unit tests only.

Closes #360

### CODE REVIEW

File: cmd/login.go [hunk 1]
```diff
@@ -526,3 +526,16 @@ func verifyLoginLegacy(ac *client.AlpaconClient, token string) {
...
+// come from the authorization server, and a person reads them off the screen.
+func printDeviceCodePrompt(w io.Writer, verificationURI, userCode string) {
+	uri, uriAltered := utils.SanitizeTerminalBlock(verificationURI)
+	code, codeAltered := utils.SanitizeTerminalBlock(userCode)
+	if uriAltered || codeAltered {
+		utils.WarnTerminalTextAltered(w, "")
+	}
+	_, _ = fmt.Fprintf(w, "\nPlease authenticate by visiting:\n%s\n\n", utils.Blue(uri))
+	_, _ = fmt.Fprintf(w, "Verification code: %s\n\n", utils.Bold(code))
+}
```
[the remark this answers is not in the document]
Comment (jisung-02) [reply cmd/login.go]: Fixed in e37710b. You are right that the block sanitizer is the wrong tool here: it keeps `\n` so the Ansible inventory and playbook snippets stay pasteable, and a URL and a device code are single values with no legitimate newline in them.

Added `SanitizeTerminalLine` in `utils/utils.go` — `SanitizeTerminalText` plus the altered flag — and `printDeviceCodePrompt` now uses it for both values.

One deliberate difference from your suggestion: the comparison is against the raw original, not a CRLF-normalized one. `SanitizeTerminalBlock` folds CRLF first because a server that ends its lines the DOS way would otherwise report every multi-line guide value as altered. For a field that is one line by contract there is no such case — a CRLF in it is injection too, so it should report altered rather than be forgiven.

The test case is `TestPrintDeviceCodePrompt/warns_when_the_URL_carried_an_injected_line`, seen red against the old code, which produced exactly the forgery you described:

```
Please authenticate by visiting:
https://demo.alpacon.io

Verification code: FAKE-0000

Verification code: ABCD-1234
```

`TestSanitizeTerminalLine` covers the newline, CRLF, ANSI, bidi, and empty cases directly.

File: cmd/login.go [hunk 2]
```diff
@@ -131,8 +132,7 @@ with the saved target as the default. Non-interactive login requires a HOST or
 				utils.CliErrorWithExit("Device code request failed. %v", err)
 			}
 
-			fmt.Fprintf(os.Stderr, "\nPlease authenticate by visiting:\n%s\n\n", utils.Blue(deviceCode.VerificationURIComplete))
-			fmt.Fprintf(os.Stderr, "Verification code: %s\n\n", utils.Bold(deviceCode.UserCode))
+			printDeviceCodePrompt(os.Stderr, deviceCode.VerificationURIComplete, deviceCode.UserCode)
 			if !noBrowser {
 				utils.OpenBrowser(deviceCode.VerificationURIComplete)
 			}
```
[the remark this answers is not in the document]
Comment (jisung-02) [reply cmd/login.go]: Fixed in a61e9ff. Good catch — this was the same substitution the sanitizing was added to prevent, and the warning above the prompt made it worse by telling the operator the value had been cleaned while the browser still received the original.

`printDeviceCodePrompt` now returns the URL it printed, and the call site opens that:

```go
promptURI := printDeviceCodePrompt(os.Stderr, deviceCode.VerificationURIComplete, deviceCode.UserCode)
if !noBrowser {
	utils.OpenBrowser(promptURI)
}
```

Returning it rather than sanitizing again at the call site leaves one value and no way to reach for the raw one by accident. `TestPrintDeviceCodePrompt` now asserts the returned URL per case and that it appears in the output, so the screen and the browser are pinned to each other. Seen red against a version that returned the raw argument:

```
--- FAIL: TestPrintDeviceCodePrompt/warns_when_the_URL_carried_an_escape
    Error: Not equal:
           expected: "https://demo.alpacon.iohttps://evil.example.com"
           actual  : "https://demo.alpacon.io\x1b[2Khttps://evil.example.com"
```

On the OS-level opener: the four other `OpenBrowser` call sites — `api/mfa/mfa.go` (two), `api/event/sudolistener.go`, and the two workspace update commands — print their URL raw and open that same raw value, so none of them has a mismatch. They are unsanitized output paths in their own right, but they are untouched by this branch and are a separate fix rather than something to fold in here.

**Review by geunwoonoh — APPROVED** [review]
## Summary

This PR fixes a security issue where guide commands and device flow authentication codes are displayed to the operator without sanitization. Control characters from the server response could be used to forge fake commands or prompt lines. The fix adds terminal text sanitization to both code paths and warns when control characters are removed.

**Test results**: ✓ `go test -race -v ./...` PASS | ✓ `go vet ./...` clean

---

## Go Convention Compliance

✓ **Error handling** - Correct use of `_, _ =` pattern for intentionally discarded errors in `fmt.Fprintf` calls

✓ **Function naming** - Follows CamelCase conventions properly

✓ **Import organization** - Added `io` import correctly where needed

✓ **Comments style** - Explains non-obvious invariants well (CRLF folding, line vs block, why return printed URI)

✓ **Test patterns** - Table-driven tests with testify/assert assertions, matching project conventions

✓ **Return values** - Named return `printedURI` in `printDeviceCodePrompt` makes the intent clear

---

## Code Quality Analysis

✓ **No Duplicated Code/Logic** - `printGuideCommand` and `printDeviceCodePrompt` intentionally differ in sanitization strategy, color handling, indent, and value count. Minimal duplication is justified by design.

✓ **No Unused Code/Functions** - All new functions are used immediately by their callers. All imports are used.

✓ **No Performance Issues** - Guide/prompt operations are infrequent (during setup/login), not on critical path. Clarity > micro-optimization is correct trade-off. CRLF handling is efficient.

---

## Security & Design Correctness

✓ **Sanitization then coloring** - Sanitizes before applying `Blue()` / `Bold()` wrappers. Reverse order would strip highlights along with escapes.

✓ **Return the sanitized URL** - `printDeviceCodePrompt` returns the URL it printed. Browser opens the URL the operator read, not the raw response. Prevents a substitution attack.

✓ **Line vs block sanitization strategy** - Device flow uses line sanitizer (newlines are injection), guide commands use block sanitizer (newlines are legitimate content). Each contract is correctly reflected.

✓ **Warnings are not silent** - Altered values are announced. Correct trade-off: altered display field is cosmetic; altered command is execution.

✓ **Comprehensive test coverage** - Exhaustive field coverage catches new fields added to response later. Tests demonstrate each failure case against the original bug.

---

## Edge Cases Handled Well

- CRLF folding: DOS line endings don't trigger false warnings
- Lone `\r`: Still reported as altered (cursor overwrite attack)
- Empty strings: Handled without error
- Tabs and control characters: Stripped correctly
- Bidi overrides: Detected and removed
- Multi-line preservation: Block sanitizer keeps newlines intact for pasteable content

---

## Checklist

- [x] Code follows Go conventions (naming, imports, error handling, comments)
- [x] No code duplication beyond intentional design
- [x] No unused code or functions
- [x] No performance degradation (infrequent operations, small scale)
- [x] Security fix is correctly implemented (sanitize→warn→print→open-sanitized)
- [x] Tests included and all passing
- [x] No security vulnerabilities introduced
- [x] Edge cases covered (CRLF, lone CR, empty strings, special characters)
- [x] CLAUDE.md conventions followed (error handling, test patterns, comments)

---

**Verdict**: APPROVED ✓

This is high-quality security work. The design is sound, the tests are thorough, and the code follows all Go and project conventions. The fix correctly addresses the threat model (operator reading forged commands) without over-sanitizing (warnings instead of silent removal, returning sanitized URL instead of raw).
